### ISSUE TITLE
Possible memory leak - tuples are not freed after mass deletion

### ISSUE THREAD

**Issue by alyapunov** [tarantool/tarantool#6078]
Tarantool 2.4.3.0-g5180d98f1

There was a strange situation. There was a common datagrid cluster - (v)shard of replica sets. There were several spaces with millions of tuples, and among them there was a space with rather huge tuples and it consumed the major part of memory. That space was partially cleaned by some internal clean-up lua procedure (removal of old versions).
After that `space:count()` was decreased significantly on every node, and on the majority of nodes the memory consumption decreased correspondingly. But on several nodes (about 8% of them) tuples were deleted from space but not actually deleted from arena. That can be seen in `box.slab.stats()` - item_count is significantly greater that corresponding `space:count()`. All those 'bad' nodes were from different replicasets - that means that there were replicasets where on replaca consumed much more memory than the others.

I see only two reasons of that: 1) lua application problem: many tuples were referenced by lua (or luaC) 2) tarantool core problem: it didn't unreferenced that many tuples (after removing from space).

We had little time to dig into problem, then the servers were restarted and the problem disappeared. We examined only that huge space - I don't know what's the situation with other spaces. Of course we tried `collectgarbage('collect')`, it didn't help. We also tried to run a script that recursively counts number of global lua variable that points to tuples - zero found. But it proves nothing - there can be local variables which references tuples.

<details>
<summary>Some raw numbers:</summary>

That is not actual output, I reconstructed values from screenshots.
```lua
-- Bad replica
tarantool> space:count()
---
- 575380
...
tarantool> space:bsize()
---
- 33437555260
...
tarantool> box.slab.stats()[#box.slab.stats()-1].item_count
---
- 1038270
...

-- Good replica
tarantool> space:count()
---
- 575380
...
tarantool> space:bsize()
---
- 33437555260
...
tarantool> box.slab.stats()[#box.slab.stats()-1].item_count
---
- 573338
...
```
</details>

I bet #6072 would help to differentiate application lua problem from tarantool core problem if we met the it again.

**Comment by alyapunov** [comment]
I think that all we can do here - is try to reproduce.

**Comment by d1rebear** [comment]
Looks like we are expiriencing something kind of similar with tarantool 2.8.3.

**Comment by alyapunov** [comment]
We are unable to waste time for random scenario generation.
Will be fixed when we fix ASAN #7326 #7327
